File: python/day-14.py
class Difference:
    def __init__(self, a):
        self.__elements = a
        self.maximumDifference = 0
        self.__elements.sort()

    def computeDifference(self):
        self.maximumDifference = self.__elements[-1] - self.__elements[0]


class Difference2:
    def __init__(self, a):
        self.__elements = a
        self.maximumDifference = 0
        self.minNum = a[0]
        self.maxNum = a[0]

        for num in self.__elements:
            self.minNum = min(self.minNum, num)
            self.maxNum = max(self.maxNum, num)

    def computeDifference(self):
        self.maximumDifference = self.maxNum - self.minNum


class Difference3:
    def __init__(self, a):
        self.__elements = a
        self.maximumDifference = 0

    def computeDifference(self):
        for i in range(len(self.__elements) - 1):
            for j in range(i + 1, len(self.__elements)):
                currDiff = abs(self.__elements[i] - self.__elements[j])
                if currDiff > self.maximumDifference:
                    self.maximumDifference = currDiff


"""
Sample Input
STDIN   Function
-----   --------
3       __elements[] size N = 3
1 2 5   __elements = [1, 2, 5]


Sample Output
4
"""
_ = input()
a = [int(e) for e in input().split(' ')]

d = Difference(a)
d.computeDifference()

print(d.maximumDifference)

File: go/day-14.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

type Difference struct {
	elements          []int32
	MaximunDifference int32
}

func (d *Difference) computeDifference() {
	for i := 0; i < len(d.elements)-1; i++ {
		for j := i + 1; j < len(d.elements); j++ {
			currDiff := Abs(d.elements[i] - d.elements[j])
			if currDiff > d.MaximunDifference {
				d.MaximunDifference = currDiff
			}

		}
	}
}

func Abs(x int32) int32 {
	if x < 0 {
		return -x
	}
	return x
}

func main() {
	/*
				Sample Input
				STDIN   Function
				-----   --------
				3       __elements[] size N = 3
				1 2 5   __elements = [1, 2, 5]

				Sample Output
				4

		3
		1 2 5
	*/

	reader := bufio.NewReaderSize(os.Stdin, 16*1024*1024)

	nTemp, err := strconv.ParseInt(strings.TrimSpace(readLine(reader)), 10, 64)

	checkError(err)
	n := int32(nTemp)

	arrTemp := strings.Split(strings.TrimSpace(readLine(reader)), " ")

	var arr []int32
	for i := 0; i < int(n); i++ {
		arrItemTemp, err := strconv.ParseInt(arrTemp[i], 10, 64)
		checkError(err)
		arrItem := int32(arrItemTemp)
		arr = append(arr, arrItem)
	}

	d := &Difference{elements: arr, MaximunDifference: 0}
	d.computeDifference()

	fmt.Println(d.MaximunDifference)
}

func readLine(reader *bufio.Reader) string {
	str, _, err := reader.ReadLine()
	if err == io.EOF {
		return ""
	}

	return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}
